File: app/src/main/java/com/crossbowffs/quotelock/account/syncadapter/SyncAdapter.java
package com.crossbowffs.quotelock.account.syncadapter;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.AbstractThreadedSyncAdapter;
import android.content.ContentProviderClient;
import android.content.Context;
import android.content.SyncResult;
import android.os.Bundle;
import android.text.TextUtils;

import androidx.core.util.Pair;

import com.crossbowffs.quotelock.BuildConfig;
import com.crossbowffs.quotelock.backup.RemoteBackup;
import com.crossbowffs.quotelock.collections.provider.QuoteCollectionHelper;
import com.crossbowffs.quotelock.utils.Xlog;

import java.util.Objects;

/**
 * @author Yubyf
 * @date 2021/6/20.
 */
public class SyncAdapter extends AbstractThreadedSyncAdapter {

    private static final String TAG = SyncAdapter.class.getSimpleName();
    public static final String SYNC_MARKER_KEY = BuildConfig.APPLICATION_ID + ".sync.marker";
    public static final String SYNC_TIMESTAMP_KEY = BuildConfig.APPLICATION_ID + ".sync.timestamp";

    private final AccountManager mAccountManager;

    private final Context mContext;

    public SyncAdapter(Context context, boolean autoInitialize) {
        super(context, autoInitialize);
        mContext = context;
        mAccountManager = AccountManager.get(context);
    }

    @Override
    public void onPerformSync(Account account, Bundle extras, String authority,
                              ContentProviderClient provider, SyncResult syncResult) {
        Xlog.d(TAG, "Performing account sync...");
        if (!RemoteBackup.Companion.getInstance().isGoogleAccountSignedIn(mContext)) {
            Xlog.d(TAG, "No account signed in.");
            syncResult.stats.numAuthExceptions++;
            return;
        }
        RemoteBackup.Result result;
        int action = checkBackupOrRestore(account);
        if (action == 0) {
            return;
        } else if (action < 0) {
            Xlog.d(TAG, "Performing remote restore...");
            result = RemoteBackup.Companion.getInstance().performSafeDriveRestoreSync(mContext,
                    QuoteCollectionHelper.DATABASE_NAME);
        } else {
            Xlog.d(TAG, "Performing remote backup...");
            result = RemoteBackup.Companion.getInstance().performSafeDriveBackupSync(mContext,
                    QuoteCollectionHelper.DATABASE_NAME);
        }
        if (result.getSuccess()) {
            Xlog.d(TAG, "Account sync success");
            syncResult.stats.numInserts++;
            setServerSyncMarker(account, result.getMd5(), result.getTimestamp());
        } else {
            Xlog.d(TAG, "Account sync failed");
            syncResult.stats.numIoExceptions++;
        }
    }

    /**
     * Check whether the current sync action should be backup, restore or nothing.
     *
     * @return the value {@code 0} if nothing to do;
     * a value less than {@code 0} if should be restore; and
     * a value greater than {@code 0} if should be backup
     */
    private int checkBackupOrRestore(Account account) {
        String serverMarker = getServerSyncMarker(account);
        long syncTimestamp = getSyncTimestamp(account);
        Pair<String, Long> databaseInfo = RemoteBackup.Companion.getInstance()
                .getDatabaseInfo(mContext, QuoteCollectionHelper.DATABASE_NAME);
        int result = 0;
        if (TextUtils.isEmpty(serverMarker) || syncTimestamp < 0
                || TextUtils.isEmpty(databaseInfo.first)) {
            Xlog.d(TAG, "First sync or local database is not created, need to perform remote restore");
            result = -1;
        } else {
            if (Objects.equals(databaseInfo.first, serverMarker)) {
                Xlog.d(TAG, "Database not changed, no need to sync");
            } else if (databaseInfo.second != null && databaseInfo.second > syncTimestamp) {
                result = 1;
            } else {
                result = -1;
            }
        }
        return result;
    }

    /**
     * This helper function fetches the last known marker
     * we received from the server - or null if we've never synced.
     */
    private String getServerSyncMarker(Account account) {
        return mAccountManager.getUserData(account, SYNC_MARKER_KEY);
    }

    private long getSyncTimestamp(Account account) {
        String timestampString = mAccountManager.getUserData(account, SYNC_TIMESTAMP_KEY);
        if (!TextUtils.isEmpty(timestampString)) {
            return Long.parseLong(timestampString);
        }
        return -1;
    }

    /**
     * Save off the last sync marker and timestamp from the server.
     */
    private void setServerSyncMarker(Account account, String marker, long timestamp) {
        mAccountManager.setUserData(account, SYNC_MARKER_KEY, marker);
        mAccountManager.setUserData(account, SYNC_TIMESTAMP_KEY, Long.toString(timestamp));
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/account/syncadapter/SyncAdapter.kt
package com.crossbowffs.quotelock.account.syncadapter

import android.accounts.Account
import android.accounts.AccountManager
import android.content.AbstractThreadedSyncAdapter
import android.content.ContentProviderClient
import android.content.Context
import android.content.SyncResult
import android.os.Bundle
import android.text.TextUtils
import com.crossbowffs.quotelock.account.google.GoogleAccountHelper
import com.crossbowffs.quotelock.data.modules.collections.QuoteCollectionRepository
import com.crossbowffs.quotelock.data.modules.collections.database.QuoteCollectionContract
import com.crossbowffs.quotelock.utils.Xlog
import com.crossbowffs.quotelock.utils.className
import com.crossbowffs.quotelock.utils.getDatabaseInfo
import com.yubyf.quotelockx.BuildConfig
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent
import kotlinx.coroutines.runBlocking

/**
 * @author Yubyf
 * @date 2021/6/20.
 */
class SyncAdapter(private val mContext: Context, autoInitialize: Boolean) :
    AbstractThreadedSyncAdapter(mContext, autoInitialize) {

    private val mAccountManager: AccountManager = AccountManager.get(mContext)

    override fun onPerformSync(
        account: Account, extras: Bundle, authority: String,
        provider: ContentProviderClient, syncResult: SyncResult,
    ) {
        Xlog.d(TAG, "Performing account sync...")
        if (!GoogleAccountHelper.isGoogleAccountSignedIn(mContext)) {
            Xlog.d(TAG, "No account signed in.")
            syncResult.stats.numAuthExceptions++
            return
        }
        val repository = EntryPointAccessors.fromApplication(
            mContext.applicationContext, SyncAdapterEntryPoint::class.java).collectionRepository()
        val action = checkBackupOrRestore(account)
        val result = when {
            action == null -> {
                Xlog.d(TAG, "No backup or restore action found. Retry 5 minutes later.")
                syncResult.delayUntil = (System.currentTimeMillis() / 1000) + 5 * 60
                return
            }
            action == 0 -> {
                Xlog.d(TAG, "Database not changed, no need to sync")
                return
            }
            action < 0 -> {
                Xlog.d(TAG, "Performing remote restore...")
                runBlocking { repository.gDriveRestoreSync() }
            }
            else -> {
                Xlog.d(TAG, "Performing remote backup...")
                runBlocking { repository.gDriveBackupSync() }
            }
        }
        if (result.success) {
            Xlog.d(TAG, "Account sync success")
            syncResult.stats.numInserts++
            syncResult.fullSyncRequested
            mAccountManager.setServerSyncMarker(account, result.md5, result.timestamp)
        } else {
            Xlog.d(TAG, "Account sync failed")
            syncResult.stats.numIoExceptions++
        }
    }

    /**
     * Check whether the current sync action should be backup, restore or nothing.
     *
     * @return the value `0` if nothing to do;
     * a value less than `0` if should be restore; and
     * a value greater than `0` if should be backup
     */
    private fun checkBackupOrRestore(account: Account): Int? {
        val serverMarker = mAccountManager.getServerSyncMarker(account)
        val syncTimestamp = mAccountManager.getSyncTimestamp(account)
        val databaseInfo = mContext.getDatabaseInfo(QuoteCollectionContract.DATABASE_NAME)
        return if (serverMarker.isNullOrEmpty() || syncTimestamp < 0 || databaseInfo.first.isNullOrEmpty()
        ) {
            Xlog.d(TAG,
                "First sync or local database is not created, need to perform remote restore")
            null
        } else {
            when {
                databaseInfo.first == serverMarker -> 0
                databaseInfo.second != null && databaseInfo.second!! > syncTimestamp -> 1
                else -> -1
            }
        }
    }

    companion object {
        private val TAG = className<SyncAdapter>()
        const val SYNC_MARKER_KEY = BuildConfig.APPLICATION_ID + ".sync.marker"
        const val SYNC_TIMESTAMP_KEY = BuildConfig.APPLICATION_ID + ".sync.timestamp"
    }

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface SyncAdapterEntryPoint {
        fun collectionRepository(): QuoteCollectionRepository
    }
}

/**
 * This helper function fetches the last known marker
 * we received from the server - or null if we've never synced.
 */
fun AccountManager.getServerSyncMarker(account: Account): String? {
    return getUserData(account, SyncAdapter.SYNC_MARKER_KEY)
}

fun AccountManager.getSyncTimestamp(account: Account): Long {
    val timestampString = getUserData(account, SyncAdapter.SYNC_TIMESTAMP_KEY)
    return if (!TextUtils.isEmpty(timestampString)) {
        timestampString.toLong()
    } else -1
}

/**
 * Save off the last sync marker and timestamp from the server.
 */
fun AccountManager.setServerSyncMarker(account: Account, marker: String?, timestamp: Long) {
    setUserData(account, SyncAdapter.SYNC_MARKER_KEY, marker)
    setUserData(account, SyncAdapter.SYNC_TIMESTAMP_KEY, timestamp.toString())
}